### ISSUE TITLE
From constrained set and min/max for common types

### ISSUE THREAD

**Issue by chaitanyagurrapu** [AutoFixture/AutoFixture#476]
It seems that if I want to limit the length of strings the recommended way is to use a custom specimen builder as discussed here http://stackoverflow.com/questions/10125199/autofixture-configure-fixture-to-limit-string-generation-length

It would be nice if Autofixture came with some of this constraints out of the box for common types such as strings, ints, datetimes, etc. So that users don't have to write them themselves.

It would also be very convenient if there was a way to specify a constrained list (of IEnumerable) for a particular property. For example, in Australia bank accounts have a BSB number, which is basically a branch number of a particular bank.

```c#
var allowedBSBs = new [] { "061237", "061238", "061239"};
var newAccount = fixture.Build<Account>()
    .With(x => x.BSB, allowedBSBs);
```

or alternate, more expressive syntax

```c#
var allowedBSBs = new [] { "061237", "061238", "061239"};
var newAccount = fixture.Build<Account>()
    .With(x => x.BSB)
    .FromConstrainedSet(allowedBSBs);
```

**Comment by ploeh** [comment]
AutoFixture was originally conceived as a TDD tool, and that's still a guiding principle behind its design. I don't mind if people use it for other purposes as well, but TDD is the main focus.

TDD is a feedback mechanism; as [GOOS](http://bit.ly/growingoos) explains, you should *listen to your tests*. AutoFixture tends to amplify this feedback mechanism, and that's one reason I find it valuable.

Whenever I encounter the need to slice and dice primitive values in various ways, I consider it a sign that I should pause and reflect. Often, the solution isn't to introduce special ways to slice and dice primitives, but instead to [refactor from primitives to Domain Objects](http://blog.ploeh.dk/2015/01/19/from-primitive-obsession-to-domain-modelling). When you do that, not only does your domain model become much clearer, but it also [becomes easier to tweak AutoFixture in your desired direction](http://blog.ploeh.dk/2009/05/01/DealingWithConstrainedInput).

Instead of treating e.g. *BSB number* as a string, make it a [Value Object](http://martinfowler.com/bliki/ValueObject.html) and then configure AutoFixture to automatically generate values drawn from `{ "061237", "061238", "061239"}` whenever a request is made for a `BsbNumber` value.

**Comment by chaitanyagurrapu** [comment]
But even if BSB is a VO rather than a string, to use Autofixture without building a custom specimen builder, I would have to have some way of constraining the list of allowed BSBs. It doesn't make sense to make them enums because they are system configurable as banks open/close branches. 

Same thing with length of fixed term deposits for example. It might be 3, 6 or 12 months for example. But in the domain the banks would like to offer different types of financial products with different maturity periods, so it doesn't make sense to make them enums either. 

Yet another example is credit card numbers. These have to pass a checksum algorithm. A custom specimen builder might be a good solution, but it would be so much easier to start with a white list first. Then I can just ask for a user with bunch of credit cards.

**Comment by ploeh** [comment]
It might make sense to add an `ElementsBuilder` which would work similarly to [FsCheck's](http://fscheck.github.io/FsCheck) `Gen.elements` or [QuickCheck's `elements`](https://hackage.haskell.org/package/QuickCheck-2.7.2/docs/Test-QuickCheck-Gen.html#v:elements) functions.

Such an `ISpecimenBuilder` should be created with a collection of elements to choose from. Perhaps it should be generic, so that an `ElementsBuilder<int>` is configured differently from an `ElementsBuilder<string>`.

**Comment by Pvlerick** [comment]
I could use this feature too so I'm willing to give it a shot if nobody is working on it.

**Comment by ploeh** [comment]
@Pvlerick I don't that anyone is :smile:

**Comment by ploeh** [comment]
Resolved in [AutoFixture 3.38.0](https://www.nuget.org/packages/AutoFixture/3.38.0).
